src/rendering/font.rs

```diff
@@ -1,195 +1,303 @@
-        if glyph.position().x + glyph.unpositioned().h_metrics().advance_width > max_x {
-            if positioned_ellipsis.is_none() {
-                max_x -= ellipsis_width;
+    pub fn right_aligned<'a>(
+        &'a self,
+        cursor: &'a mut Cursor,
+    ) -> impl Iterator<Item = PositionedGlyph> + 'a {
+        let scale = self.font.scale;
+
+        Iterator::zip(
+            self.buffer.glyph_infos().iter(),
+            self.buffer.glyph_positions().iter(),
+        )
+        .rev()
+        .map(move |(i, p)| {
+            cursor.x -= p.x_advance as f32 * scale;
+            cursor.y -= p.y_advance as f32 * scale;
+            PositionedGlyph {
+                id: i.codepoint,
+                x: cursor.x + p.x_offset as f32 * scale,
+                y: cursor.y + p.y_offset as f32 * scale,
             }
-            positioned_ellipsis = Some(ellipsis.clone().positioned(glyph.position()));
-            glyphs.pop();
-        } else {
-            break;
-        }
+        })
     }
-    if let Some(ellipsis) = positioned_ellipsis {
-        glyphs.push(ellipsis);
+
+    pub fn centered<'a>(
+        &'a self,
+        cursor: &'a mut Cursor,
+        min_x: f32,
+        max_x: f32,
+    ) -> impl Iterator<Item = PositionedGlyph> + 'a {
+        let (mut adv_x, mut adv_y) = (0, 0);
+
+        for p in self.buffer.glyph_positions() {
+            adv_x += p.x_advance;
+            adv_y += p.y_advance;
+        }
+
+        cursor.x -= adv_x as f32 * self.font.scale / 2.0;
+        cursor.y -= adv_y as f32 * self.font.scale / 2.0;
+
+        if cursor.x < min_x {
+            cursor.x = min_x;
+        }
+
+        self.left_aligned(cursor, max_x)
     }
 
-    glyphs.into_iter()
-}
+    // This is a right aligned layout where all the digits have the same width.
+    pub fn tabular_numbers<'a>(
+        &'a self,
+        cursor: &'a mut Cursor,
+    ) -> impl Iterator<Item = PositionedGlyph> + 'a {
+        let mut digits = [0; 10];
+        let mut digit_width = 0.0;
+        for (digit, glyph) in digits.iter_mut().enumerate() {
+            let GlyphId(the_glyph) = self
+                .font
+                .font
+                .face
+                .glyph_index(char::from(digit as u8 + b'0'))
+                .unwrap_or_default();
 
-pub fn measure<'fd>(layout: impl IntoIterator<Item = PositionedGlyph<'fd>>) -> f32 {
-    let mut first = None;
-    layout
-        .into_iter()
-        .inspect(|g| {
-            first.get_or_insert_with(|| g.position().x);
-        })
-        .last()
-        .map_or(0.0, |g| {
-            g.position().x + g.unpositioned().h_metrics().advance_width - first.unwrap()
+            *glyph = the_glyph as _;
+
+            let width = self.font.glyph_x_advance(the_glyph as _);
+            if width > digit_width {
+                digit_width = width;
+            }
+        }
+
+        // FIXME: There's kerning between e.g. ".1" now, which is maybe not quite
```

Comment:
Do you have an example of what this looks like?

Reply:
This is before I switched to tabular nums, but you can kind of see it here how .2 and .1 are differently aligned: 

![https://i.imgur.com/K20g1XW.png](https://i.imgur.com/K20g1XW.png)

Reply:
If the font (which you can't change yet) doesn't have tnum support, then this will still happen.

Reply:
Kerning could be turned off, but we currently have tnum active for all the values of the key value components, even if they are "info text" components. So turning kerning off for components like Current Comparison would be suboptimal. Though we could introduce a flag in the key value component state that marks it as a "info time" as opposed to a "info text" component.